### ISSUE TITLE
Distance matrix when crossing streets is not allowed

### ISSUE THREAD

**Issue by maknotavailable** [Project-OSRM/osrm-backend#5912]
I am looking to build a large distance matrix with the following restrictions:
- locations are buildings (so on either side of a street)
- distance calculated as the distance you can legally travel without crossing the street (profile = car)

Has anyone tried this with the current OSRM-Backend version? 

If not, I'd look to work on such functionality and would be greatfull for any ideas/support/..

Thanks!

**Comment by sergn-n** [comment]
Looks like you're out of luck . A number of discussions regarding legal crossing streets took place at OSM community, see [this one](https://lists.openstreetmap.org/pipermail/tagging/2019-October/048646.html) for example. Currently, this feature is missing that is no common solution. OSRM just has no reliable data to extract from OSM to know if crossing the street is [il]legal.

**Comment by maknotavailable** [comment]
Thanks for the reply!

My case goes a bit beyond legal/illegal crossings. I need to apply the following rules:
- if the speed limit is over 50 km/h, crossing the street is not allowed
- if a special vehicle is used (sideloader), then crossing the street is not allowed

So in both cases, the information of what side of the street a "stop" is on, and possibly the speed limit, are the only information I need to make a decision about how to navigate to those stops. This leaves me with a problem, where I am not sure how to best approach it:
- build a (large) distance matrix where I can take into account these two constraints.

**Comment by danpat** [comment]
> if the speed limit is over 50 km/h, crossing the street is not allowed

This can be calculated ahead of time.  If you know this, then OSRM's `approaches=curb` parameter could be applied to that particular waypoint.  Unfortunately, we've only implemented this for the `/route` service, not the `/table` service.

I remember having internal discussions about adding an `auto` setting for `approaches=`, using things like road speeds as heuristics to determine whether it's safe.  It was never implemented though.

> if a special vehicle is used (sideloader), then crossing the street is not allowed

Do you know this ahead of time, or is this only determined once you perform your optimisation step?  If you do not know ahead of time which vehicle type is used for each stop, then you would need *both* the "arrive on the curb side" and "arrive on the fastest side" options for your solver to pick from.  If we got `approaches=` working, then you'd simply add the same coordinate twice to the `/table` request, with `approaches=curb` for one, and `approaches=any` for the other.

So, TL;DR - you can't do this right now, the blocker is support for `approaches=` parameter on the `/table` service.

**Comment by maknotavailable** [comment]
Thanks a lot for the info!

It is known ahead of time what type of vehicle will be driving a given route. Therefore, using `approaches=curb` should solve my problem!

Any pointers for implementing this in the `/table` service? I do not have extensive experience with C++, but would be willing to look into it.
